cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+In short, the following principles should be followed to reach a secure file upload implementation:
+
+- **Whitelist allowed extensions. Only allow safe and critical extensions for business functionality**
+  - **Ensure that [input validation](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Input_Validation_Cheat_Sheet.md#validating-free-form-unicode-text) is applied before validating the extensions.**
+- **Allowed users only can upload files**
+- **Validate the file type, don't trust the [Content-Type header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) as it can be spoofed**
+- **Change the file name to something generated by the application**
+- **Set a file name length limit. Restrict the allowed characters if possible**
+  - **Setting a character limit could deny the attacker from sending null bytes, adding secondary extensions, and performing XSS attacks if the file name is reflected**
+- **Set a file size limit**
+- **Store the files on a different server. If that's not possible, store them in a private directory**
+  - **In the case of public access to the files, ensure that proper permissions are set, file names are safe, and encoded**
+- **Run the file through an antivirus or a sandbox if available to validate that it doesn't contain malicious or bogus data**
+- **Ensure the usage of secure and updated libraries for your needs**
+- **Add [CSRF protection](Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md), as this is input being served from the user**
+
+## Contents
+
+- [File Upload Threats](#file-upload-threats)
+  - [Malicious Files](#malicious-files)
+  - [Public File Retrieval](#public-file-retrieval)
+- [File Upload Protection](#file-upload-protection)
+  - [Extension Validation](#extension-validation)
+     - [Whitelist Extensions](#whitelist-extensions)
+     - [Blacklist Extensions](#blacklist-extensions)
+  - [Content-Type Validation](#content-type-validation)
+  - [Magic Number Validation](#magic-number-validation)
+  - [File Name Sanitization](#file-name-sanitization)
+  - [File Content Validation](#file-content-validation)
+  - [File Storage Location](#file-storage-location)
+  - [User Permissions](#user-permissions)
+  - [System Permissions](#system-permissions)
+  - [Upload and Download Limits](#upload-and-download-limits)
+- [Java Code Snippets](#java-code-snippets)
+
+## File Upload Threats
+
+In order to assess and know exactly what controls to implement, knowing what you're facing is essential to protect your assets. The following sections will hopefully showcase the risks accompanying the file upload functionality.
+
+### Malicious Files
+
+The attacker delivers a file for malicious intent, such as:
+
+1. Exploit vulnerabilities in the file parser or processing module (_e.g._ [ImageTrick Exploit](https://imagetragick.com/), [XXE](https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Processing))
+2. Use the file for phishing (_e.g._ careers form)
```

Comment:
I'm not entirely sure what you're talking about here. Would this be a CV with macros to try and exploit a user (covered in point 5), or a HTML page you upload to steal creds, or a CV with a link to a dodgy site?

Reply:
Point 5 doesn't cover phishing attacks using macros. I will update point 5 to be clearer